karate-core/src/test/java/com/intuit/karate/MatchTest.java

```diff
@@ -379,4 +379,12 @@ void testOptional() {
         match("{ a: null}", EQUALS, " { a: '##notnull' }");
     }
 
+    @Test
+    void testEmbedded() {
+        assertAll(
+            () -> match("[]", CONTAINS, "#[] ##({ a: \"#number\" })"),
```

Comment:
this does not make sense to me. `#[]` is for an array. and if you are trying to substitute a variable you just do `'#[] varName'`

what is the use of this, at least I never intentionally designed for this

Reply:
to make it easier to discuss, let me start with this example:

```cucumber
* def chunk = { foo: 'bar' }
* def schema = '#[] chunk'
* match [] == schema
```

so what more are you trying to support ?

Reply:
The simplest explanation is reusability. I have a schema defined in a json file, which is read in the tests (there is a reason for this but explaining it will just add more complexity). With that, when I make a request to an endpoint if no data is available an empty array is returned in the response. If data is available the response contains an array of objects. In order to use the same schema for both conditions it needs to match an empty array or an array with data and also verify that the data types match what is defined in the schema. So "#[]" matches an array, "#(schema)" matches the data. "#[] ##(schema)" will match either an empty array, or if the array contains objects it will match the data types defined in the schema to the objects returned.

Reply:
`#[]` is the same as `match each`. the way this has always worked in karate is that it will match an empty array as well as a non-empty array.

so let me revise my example:

```cucumber
* def chunk = { foo: '#string' }
* def schema = '#[] chunk'
* match [] == schema
* match [{ foo: 'bar' }] == schema
```

so what more do you need ?

Reply:
The schema is read from a file... "chunk" is replaced by javascript call to read another file, which is only executed when inside `#()` as `"#(karate.read('some-other-schema.json'))"` so the "schema";  

response.json (not really valid json but it works)
```
"#[] ##(karate.read('some-other-schema.json'))"
```

and the test now has a one liner match making it more readable;
... endpoint call here
* match response contains read('response.json')

Reply:
> not really valid json but it works

I don't like this at all, and I personally would hate to be someone who has to maintain your tests, just my opinion. anyway I really don't want to make Karate support this for that reason.

I was trying to track down some other issue, but am back. so let me try again with the latest version from develop. this works:

```cucumber
* def schema = "#[] read('schema-read.json')"
* match [] == schema
* match [{ foo: 'bar' }] == schema
```

and `schema-read.json` is `{ "foo": "#string" }` - and I think you have everything you need.

maybe I'm dense but I still don't understand the mega-indirection of your above example. so can you please stick to my example and extend it ? thanks. and also maybe you should read this: https://stackoverflow.com/a/54126724/143475

Reply:
Well that just happened to be a simple example, where the json was invalid. Another, valid scenario is

schema.json
```
{
  "id": "#string",
  "items": "#[] ##(karete.read('item.json'))"
}
```

which is valid json, with the embedded expression match, for the items array that contains json objects.

Reply:
I really don't understand (and don't like) the use of a double-hash, and I don't think it is needed.

I just made a commit, here's a screenshot below: https://github.com/intuit/karate/commit/f9dcfe0f56db15efd386ce4ad0f68a6f3bc73134

so please tell me now whether it does the same thing you want or not.

<img width="485" alt="image" src="https://user-images.githubusercontent.com/915480/129225788-97983d2f-095a-48ae-bd7a-1a44ce043549.png">

Reply:
Yup, looks like that example does work as well in 1.0.1 without the double-hash, however, it does not work in 1.1.0.

Reply:
it does not work because using raw JSON (e.g. `[]`) on the LHS was broken, if you use variables it will work.
but anyway, can I please close this ticket ?

Reply:
Sure?